### PR TITLE
bpo-40619: Correctly handle error lines in programs without file mode

### PR DESCRIPTION
**Pull request by pablogsal**
https://bugs.python.org/issue40619

### CODE REVIEW

**Review by lysnikolaou — APPROVED** [review]
> +3 −25

Nice! Was it so simple all along?

**Comment by pablogsal** [comment]
> Nice! Was it so simple all along?

And one of those lines is a test and the other is to make the only line more readable 😉

**Review by gvanrossum — COMMENTED** [review]
This looked too good to be true, and it is. :-(
```
>>> compile("(1+)", "<string>", "eval")
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "<string>", line 1
    (1+
       ^
SyntaxError: unexpected EOF while parsing
>>> 
```

**Comment by pablogsal** [comment]
> This looked too good to be true, and it is. :-(

:( It seems that is the `size - 1` that is removing that character. Using `size` instead makes this test fail:

```
 check('Python = "\u1e54\xfd\u0163\u0125\xf2\xf1" +', 1, 20)
```

**Comment by gvanrossum** [comment]
Perhaps you need `size - 1` if the character before `size` is a `'\n'` and `size` otherwise?

**Comment by gvanrossum** [comment]
Also: Think very carefully about that example you quoted. It could be a bytes vs. unicode issue and much deeper than subtracting 1 if the last character is a newline.

**Comment by pablogsal** [comment]
> Also: Think very carefully about that example you quoted. It could be a bytes vs. unicode issue and much deeper than subtracting 1 if the last character is a newline.

I think the difference that that is reporting is precisely the Unicode size of the "\n" that we append that is not on the original string.

**Comment by pablogsal** [comment]
> Perhaps you need `size - 1` if the character before `size` is a `'\n'` and `size` otherwise?

This seems to do the trick

File: Parser/pegen/pegen.c [hunk 1]
```diff
@@ -399,7 +399,9 @@ _PyPegen_raise_error_known_location(Parser *p, PyObject *errtype,
 
     if (!error_line) {
         Py_ssize_t size = p->tok->inp - p->tok->buf;
-        error_line = PyUnicode_DecodeUTF8(p->tok->buf, size ? size - 1: size, "replace");
+        error_line = PyUnicode_DecodeUTF8(p->tok->buf,
+                                          p->tok->buf[size-1] == '\n' ? size - 1 : size,
+                                          "replace");
```
Comment (gvanrossum) [inline Parser/pegen/pegen.c]: What if size is zero? I'd use an `if`:
```
if (size && p->tok->buf[size-1] == '\n')
    size--;
```

**Review by gvanrossum — APPROVED** [review]
Go for it!

**Pull request merged**